### PR TITLE
Feature/redbox 171 add user filtering to core api

### PR DESCRIPTION
**Pull request by gecBurton**
## Context

As a User I only want to be able to GET & DELETE my own files, and no one else should be able to access them

## Changes proposed in this pull request

* I have added `user_uuid` to various storage methods that search for data to ensure that it only returns data belonging to the relevant user
* In `core_api/src/routes/file.py` I have ensured that we are now using the `user_uuid` to check that the user can only access their own data
* I *have not* tackled the important issue of how to ensure that the `/chat/rag` endpoint filters data belonging to the user because:
  1. I wasn't sure how to
  2. I think that there maybe some change to this endpoint anyway to allow for streaming 
  3. This PR is big enough

## Guidance to review

All fairly dull stuff tbh. I have added a lot of tests, they should pass 🤞 

## Relevant links

https://technologyprogramme.atlassian.net/browse/REDBOX-171

## Things to check

- [ ] I have added any new ENV vars in all deployed environments
- [ ] I have tested any code added or changed
- [x] I have run integration tests https://github.com/i-dot-ai/redbox-copilot/actions/runs/8967112884

### CODE REVIEW

File: core_api/tests/conftest.py [hunk 1]
```diff
@@ -19,11 +18,6 @@
 YieldFixture = Generator[T, None, None]
 
 
-@pytest.fixture(autouse=True)
```
Comment (gecBurton) [inline core_api/tests/conftest.py:22]: not used

File: core_api/tests/test_auth.py [hunk 2]
```diff
@@ -11,7 +11,7 @@
         ({"Authorization": "blah blah"}, 403),
         ({"Authorization": "Bearer blah-blah"}, 401),
         ({"Authorization": "Bearer " + jwt.encode({"user_uuid": "not a uuid"}, key="super-secure-private-key")}, 401),
-        ({"Authorization": "Bearer " + jwt.encode({"user_uuid": str(uuid4())}, key="super-secure-private-key")}, 200),
+        ({"Authorization": "Bearer " + jwt.encode({"user_uuid": str(uuid4())}, key="super-secure-private-key")}, 404),
```
Comment (gecBurton) [inline core_api/tests/test_auth.py:14]: this now has to be right right uuid not just any one

File: redbox/export/docx.py [hunk 3]
```diff
@@ -56,7 +56,7 @@ def spotlight_complete_to_docx(
 
     document.add_heading("Summarised Files", level=1)
     for file in files:
-        document.add_paragraph(file.name, style="List Bullet")
+        document.add_paragraph(file.key, style="List Bullet")
```
Comment (gecBurton) [inline redbox/export/docx.py:59]: minor bug fixes here, is this code used?

File: redbox/models/base.py [hunk 4]
```diff
@@ -1,16 +1,15 @@
...
 
-from pydantic import AfterValidator, BaseModel, Field, computed_field
+from pydantic import BaseModel, Field, computed_field
 
 
 class PersistableModel(BaseModel):
     """Base class for all models that can be persisted to the database."""
 
-    uuid: UUID | Annotated[str, AfterValidator(lambda x: UUID(x))] = Field(default_factory=uuid4)
+    uuid: UUID = Field(default_factory=uuid4)
```
Comment (gecBurton) [inline redbox/models/base.py:10]: simpler

File: redbox/models/file.py [hunk 5]
```diff
@@ -79,7 +79,12 @@ class Metadata(BaseModel):
     # url: Optional[str] = None
 
     @classmethod
-    def merge(cls, left: Metadata, right: Metadata) -> Metadata:
+    def merge(cls, left: Optional[Metadata], right: Optional[Metadata]) -> Optional[Metadata]:
+        if not left:
```
Comment (gecBurton) [inline redbox/models/file.py:83]: mypy started complaining about this, `reduce` needs this behaviour

File: redbox/parsing/chunk_clustering.py [hunk 6]
```diff
@@ -10,8 +10,8 @@
 
 def cluster_chunks(
     chunks: list[Chunk],
+    embedding_model: SentenceTransformer,
```
Comment (gecBurton) [inline redbox/parsing/chunk_clustering.py:13]: this was never really optional

File: redbox/parsing/chunk_clustering.py [hunk 7]
```diff
@@ -127,10 +127,3 @@ def create_pdist(token_counts, pair_embed_dist, weight_embed_dist=0.2, use_log=T
     # the two above distance are combined either using sum or product (i.e. use_log=T)
     combined_dist = [x + y for x, y in zip(embed_dist, token_dist)]
     return combined_dist
-
-
-def merge_chunk_metadata(meta_in: list[Metadata]) -> Metadata:
```
Comment (gecBurton) [inline redbox/parsing/chunk_clustering.py:132]: not used

File: redbox/storage/elasticsearch.py [hunk 8]
```diff
@@ -147,30 +147,28 @@ def get_file_chunks(self, parent_file_uuid: UUID) -> list[Chunk]:
...
+                                },
+                            ]
+                        }
+                    }
+                },
             )
         ]
         return res
 
-    def _get_child_chunks(self, parent_file_uuid: UUID) -> list[UUID]:
```
Comment (gecBurton) [inline redbox/storage/elasticsearch.py:155]: duplicate of get_file_chunks

File: redbox/llm/spotlight/spotlight.py [hunk 9]
```diff
@@ -9,29 +11,33 @@
 
 # region ===== TASKS =====
 
+CREATOR_USER_UUID = UUID("aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa")
```
Comment (gecBurton) [inline redbox/llm/spotlight/spotlight.py:14]: this is a bit nasty but a price worth paying to ensure that data is always saved with a `creator_user_id`?

File: redbox/storage/elasticsearch.py [hunk 10]
```diff
@@ -138,63 +138,38 @@ def list_all_items(self, model_type: str) -> list[UUID]:
...
-                    _source=False,
-                )
-            ]
-        except NotFoundError:
-            print(f"Index {target_index} not found. Returning empty list.")
-            return []
-
-        return matched_chunk_ids
-
-    def _get_embedded_child_chunks(self, parent_file_uuid: UUID) -> list[UUID]:
```
Comment (gecBurton) [inline redbox/storage/elasticsearch.py:174]: i found a simpler way to do this

File: redbox/storage/elasticsearch.py [hunk 11]
```diff
@@ -206,39 +181,31 @@ def get_file_status(self, file_uuid: UUID) -> FileStatus:
...
-            self.read_item(file_uuid, "File")
+            file = self.read_item(file_uuid, "File")
         except NotFoundError:
             raise ValueError(f"File {file_uuid} not found")
+        if file.uuid != user_uuid:
+            raise ValueError(f"File {file_uuid} not found")
 
         # Test 2: Get the number of chunks for the file
-        chunk_uuids = self._get_child_chunks(file_uuid)
+        chunks = self.get_file_chunks(file_uuid, file.creator_user_uuid)
```
Comment (gecBurton) [inline redbox/storage/elasticsearch.py:191]: reusing some existing code here

File: redbox/storage/elasticsearch.py [hunk 12]
```diff
@@ -206,39 +181,31 @@ def get_file_status(self, file_uuid: UUID) -> FileStatus:
...
 
-        # Test 3: Get the number of embedded chunks for the file
-        embedded_chunk_uuids = self._get_embedded_child_chunks(file_uuid)
-
-        chunk_statuses = [
-            ChunkStatus(chunk_uuid=chunk_uuid, embedded=chunk_uuid in embedded_chunk_uuids)
-            for chunk_uuid in chunk_uuids
-        ]
+        # Test 3: Determine the number of embedded chunks for the file
+        chunk_statuses = [ChunkStatus(chunk_uuid=chunk.uuid, embedded=bool(chunk.embedding)) for chunk in chunks]
```
Comment (gecBurton) [inline redbox/storage/elasticsearch.py:202]: a bit simpler?

File: redbox/tests/conftest.py [hunk 13]
```diff
@@ -14,29 +14,101 @@
...
+@pytest.fixture
+def env():
+    yield Settings(django_secret_key="", postgres_password="")
 
 
 @pytest.fixture
-def chunk() -> Chunk:
-    test_chunk = Chunk(
-        parent_file_uuid=uuid4(),
+def alice():
```
Comment (gecBurton) [inline redbox/tests/conftest.py:23]: open to discussion but sometimes i find it easier to name people

File: core_api/src/routes/file.py [hunk 14]
```diff
@@ -115,11 +115,14 @@ def get_file(file_uuid: UUID, _user_uuid: Annotated[UUID, Depends(get_user_uuid)
     Returns:
         File: The file
     """
-    return storage_handler.read_item(file_uuid, model_type="File")
+    file = storage_handler.read_item(file_uuid, model_type="File")
+    if file.creator_user_uuid != user_uuid:
+        raise HTTPException(status_code=404)
```
Comment (gecBurton) [inline core_api/src/routes/file.py:120]: yes this file does exist but we dont want the requestor to know that as it isnt thiers

File: redbox/llm/spotlight/spotlight.py (hunk 9 above)
Comment (252afh) [reply redbox/llm/spotlight/spotlight.py:14]: Do we need to create 2 persistable models? One with and one without user_id?

**Review by 252afh — APPROVED** [review]

File: core_api/src/routes/file.py [hunk 15]
```diff
@@ -106,55 +107,64 @@ async def upload_file(
...
 
     Args:
-        file_uuid (str): The UUID of the file to delete
+        file_uuid (UUID): The UUID of the file to delete
+        user_uuid (UUID): The UUID of the user
 
     Returns:
         File: The file that was deleted
     """
     file = storage_handler.read_item(file_uuid, model_type="File")
```
[the remark this answers is not in the document]
Comment (gecBurton) [reply core_api/src/routes/file.py:137]: i.e. add a filter to the ES query?

File: redbox/models/base.py [hunk 16]
```diff
@@ -1,16 +1,15 @@
...
     created_datetime: datetime = Field(default_factory=datetime.utcnow)
-    creator_user_uuid: Optional[UUID | Annotated[str, AfterValidator(lambda x: UUID(x))]] = None
+    creator_user_uuid: UUID
```
[the remark this answers is not in the document]
Comment (gecBurton) [reply redbox/models/base.py:12]: or even `creator_uuid`? @lmwilkigov you get the casting vote!

File: redbox/llm/spotlight/spotlight.py (hunk 9 above)
Comment (gecBurton) [reply redbox/llm/spotlight/spotlight.py:14]: https://technologyprogramme.atlassian.net/browse/REDBOX-208

File: core_api/tests/conftest.py [hunk 17]
```diff
@@ -55,13 +49,13 @@ def app_client():
 
 
 @pytest.fixture
-def user_uuid() -> UUID:
+def alice() -> UUID:
```
[the remark this answers is not in the document]
Comment (gecBurton) [reply core_api/tests/conftest.py:52]: https://technologyprogramme.atlassian.net/browse/REDBOX-209

**Pull request merged**